VIPs/vip-7.md

```diff
@@ -0,0 +1,260 @@
+author: Chris Were <chris@verida.io>
+discussions-to: #dev-protocol on Verida Discord
+status: Draft
+type: Standards Track
+category: Core
+created: 2022-01-03
+---
+
+# Overview
+
+The selection of storage nodes occurs when a DID is first created. The storage nodes are used to store the actual DID Documents and the same nodes are typically (but not necessarily) used to store data for each application context.
+
+This VIP provides a specification for the discovery of the available storage nodes and default rules for selecting an appropriate storage node.
+
+This functionality is only required by Verida wallet implementations (ie: `Verida Vault`) and Verida clients that maintain their own wallet (ie: Server applications that maintain their own Verida network private key).
+
+# Motivation
+
+The Verida protocol requires data to be stored on multiple storage nodes to enable redundancy and maximise performance in a decentralized environment.
+
+It's necessary to provide a simple storage node discovery and selection capability into the protocol to:
+
+1. Minimize developer friction when using the protocol
+2. Minimize new user friction when creating a DID
+3. Maximising flexibility for advanced users
+4. Maximising network performance for end users
+
+# Discovery
+
+Storage nodes must register theselves with the protocol so they can be discovered by new users. These storage nodes will provide global metadata that can assist new users select the most appropriate storage nodes to use.
+
+The protocol will provide a default set of logic to determine the most appropriate storage nodes for a new user, however users will be able to override this and manually select the storage nodes they wish to use. It is, after all, their data.
+
+Self-hosted storage nodes can be created and used by advanced users. They don't need to be registered with Verida. These nodes can be shared privately amongst groups of users, but won't be discoverable by general users of the Verida network.
+
+The discovery process will be implemented in two phases.
+
+## Phase 1
+
+Verida will manually verify storage node operators and maintain a list of trusted storage nodes with accurate metadata about each storage node.
+
+## Phase 2
+
+A Verida smart contract will be deployed that permits any storage node operator to register themselves on the network.
+
+This phase is out of scope for this VIP.
+
+# Selection considerations
+
+The process of selecting appropriate storage nodes requires consideration of:
+
+- Geographic location
+- Availability
+- Performance
+- Regulatory jurisdiction
+- Node operator
+- User preference
+
+The protocol must specify a default selection process for new users. It's intended that advanced users can select their own storage nodes at any time (including selecting self-hosted storage nodes).
+
+## Geographic location
+
+The location of the storage node could consist of the following relevant metadata:
+
+- Country
+- Region (ie: Europe, South East Asia)
```

Comment:
I think this needs more specification to be useful. 

I think it needs to support both sub-country specification (eg, GCP has both Sydney and Melbourne data centres, AWS has US-East-X and US-West-X etc) and larger regions (eg, Southern Africa). 

I think it needs to be free text, and is basically informational for a user to use when doing manual filtering.

Reply:
I think we need to stick with standardization here as much as possible, especially for phase 1.

As per the specification section, the intent is to use the regions from (https://github.com/lukes/ISO-3166-Countries-with-Regional-Codes/blob/master/all/all.csv). I have updated the specification to also support sub-regions so we can use that data if required.

Note: The datacenters have lots of descriptive information that can be searched on by advanced users which; when combined with long / lat should make it very easy to pinpoint the most appropriate nodes. We could even show a map.